app/app/src/main/java/de/stustapay/stustapay/storage/InfallibleApiRequestSerializer.kt

```diff
@@ -0,0 +1,123 @@
+import de.stustapay.api.models.NewTopUp
+import de.stustapay.api.models.PaymentMethod
+import de.stustapay.stustapay.model.InfallibleApiRequest
+import de.stustapay.stustapay.model.InfallibleApiRequestKind
+import de.stustapay.stustapay.model.InfallibleApiRequests
+import de.stustapay.stustapay.proto.InfallibleApiRequestKindProto
+import de.stustapay.stustapay.proto.InfallibleApiRequestProto
+import de.stustapay.stustapay.proto.InfallibleApiRequestTicketSaleProto
+import de.stustapay.stustapay.proto.InfallibleApiRequestTopUpProto
+import de.stustapay.stustapay.proto.InfallibleApiRequestsProto
+import java.io.File
+import java.io.InputStream
+import java.io.OutputStream
+import java.util.UUID
+import javax.inject.Singleton
+
+object InfallibleApiRequestSerializer : Serializer<InfallibleApiRequests> {
+    override val defaultValue: InfallibleApiRequests
+        get() = InfallibleApiRequests(requests = mapOf())
+
+    override suspend fun readFrom(input: InputStream): InfallibleApiRequests {
+        return try {
+            InfallibleApiRequests(
+                InfallibleApiRequestsProto.parseFrom(input).requestsList.mapNotNull {
+                    val id = UUID.fromString(it.id)
+                    when (it.kind) {
+                        InfallibleApiRequestKindProto.TOP_UP -> {
+                            Pair(
+                                id, InfallibleApiRequest(
+                                    InfallibleApiRequestKind.TopUp(
+                                        NewTopUp(
+                                            amount = it.topUp.amount,
+                                            customerTagUid = BigInteger.parseString(it.topUp.customerTagUid),
+                                            paymentMethod = PaymentMethod.decode(it.topUp.paymentMethod)!!,
+                                            uuid = id,
+                                        )
+                                    )
+                                )
+                            )
+                        }
+
+                        InfallibleApiRequestKindProto.TICKET_SALE -> {
+                            Pair(
+                                id, InfallibleApiRequest(
+                                    InfallibleApiRequestKind.TicketSale(
+                                        NewTicketSale(
+                                            customerTagUids = it.ticketSale.customerTagUidsList.map {
+                                                BigInteger.parseString(it)
+                                            },
+                                            paymentMethod = PaymentMethod.decode(it.ticketSale.paymentMethod)!!,
+                                            uuid = id,
+                                        )
+                                    )
+                                )
+                            )
+                        }
+
+                        InfallibleApiRequestKindProto.UNRECOGNIZED -> null
```

Comment:
switching the types here seems rather inelegant - it should better dispatch to the type-specific data source function implemented in that datasource file.
otherwise when we decide to make more requests stored locally this will become a monster-implementation.

Reply:
I don't see how moving the serialisation implementation to the DataSource improves this. I definitely agree that this code is verbose and unwieldy, but that stems from the fact that we essentially need to completely serialise a specific subset of possible api requests. I couldn't think of a way to do this generically for all API requests, and currently believe that this is impossible. Besides, only a very small subset of API requests are actually infallible, so I think this approach is fine.